nucypher/blockchain/eth/agents.py

```diff
@@ -41,31 +45,40 @@
     PREALLOCATION_ESCROW_CONTRACT_NAME,
     STAKING_ESCROW_CONTRACT_NAME,
     STAKING_INTERFACE_CONTRACT_NAME,
-    STAKING_INTERFACE_ROUTER_CONTRACT_NAME
+    STAKING_INTERFACE_ROUTER_CONTRACT_NAME,
+    WORKLOCK_CONTRACT_NAME
 )
 from nucypher.blockchain.eth.decorators import contract_api, validate_checksum_address
 from nucypher.blockchain.eth.events import ContractEvents
 from nucypher.blockchain.eth.interfaces import BlockchainInterfaceFactory
 from nucypher.blockchain.eth.registry import AllocationRegistry, BaseContractRegistry
+from nucypher.types import (
+    Agent,
+    NuNits,
+    SubStakeInfo,
+    RawSubStakeInfo,
+    Period,
+    WorklockParameters,
+    StakerFlags,
+    StakerInfo,
+    PeriodDelta,
+    StakingEscrowParameters
+)
 from nucypher.blockchain.eth.utils import epoch_to_period
 from nucypher.crypto.api import sha256_digest
 
-# TODO: Use everywhere
-# TODO: Uses string to avoid circular import
-Agent = TypeVar('Agent', bound='EthereumContractAgent')
-
 
 class ContractAgency:
+    """Where agents live and die."""
+    
     # TODO: Enforce singleton - #1506 - Okay, actually, make this into a module
-
-    __agents = dict()
-    TRANSACTION_REGISTRY = dict()
+    __agents: Dict[Type[Agent], Agent] = dict()
 
     @classmethod
     def get_agent(cls,
                   agent_class: Type[Agent],
-                  registry: BaseContractRegistry = None,
-                  provider_uri: str = None) -> Agent:
+                  registry: Optional[BaseContractRegistry] = None,
+                  provider_uri: Optional[str] = None) -> Agent:
```

Comment:
A `TypeVar` can be used here to make types stricter (to assert that the return type == `agent_class`)

Reply:
Oh neat.  I'm still getting used to this way of thinking - What does that look like?  Thanks for looking :-)

Reply:
Hm, didn't you already use it in some other place just a few days ago? Anyway, you can define `_Agent = TypeVar('_Agent', bound=Agent)` (can be a class variable) and use it as `get_agent(cls, agent_class: Type[_Agent], ...) -> _Agent`

Reply:
in this case `Agent` is already a `TypeVar`: Agent = TypeVar('Agent', bound='EthereumContractAgent')

Reply:
Ah, I see, I guess that was the episode I remembered. My bad.